ngraph/src/ngraph/op/convolution.cpp

```diff
@@ -530,6 +596,74 @@ shared_ptr<Node>
+                                                                                                      dilation,
+                                                                                                      pad_above,
+                                                                                                      pad_below,
+                                                                                                      strides);
+            return true;
+        } else {
+            return false;
+        }
+
+    }
+
+    bool evaluate_convolution_backprop_data(const HostTensorPtr &arg, const HostTensorPtr &out, const void *filter,
+                                            const Shape &filter_shape, const Strides &strides, const Strides &dilation,
+                                            const CoordinateDiff &pad_above, const CoordinateDiff &pad_below) {
+// TODO: Arithmetic operators are not defined for bfloat16. Issue 33808
+// try_evaluate_conv_bp><element::Type_t::bf16>(arg, out, filter, filter_shape, strides, dilation, pad_above, pad_below);
+        return try_evaluate_conv_bp<element::Type_t::f16>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                           pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::f32>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                           pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::i8>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                          pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::i16>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                           pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::i32>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                           pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::i64>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                           pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::u8>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                          pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::u16>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                           pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::u32>(arg, out, filter, filter_shape, strides, dilation, pad_above,
+                                                           pad_below) ||
+                try_evaluate_conv_bp<element::Type_t::u64>(arg, out, filter, filter_shape, strides, dilation, pad_above,
```

Comment:
I really don't like the idea to add any template reference implementation to nGraph library. It can have a big impact on binary size. If you add the reference implementation only for interpreter backend and it isn't used in the real topologies, please add this implementation only for nGraph tests.

For example: move header with template implementation (try_evaluate_conv_bp in this case) to `ngraph/test/runtime/reference` folder, and after that you can use this implementation in the Interpreter backend, you don't impact library size and moreover if in future we want to add this reference implementation to nGraph operation it is really easy to do.

Reply:
Please have a look at the solution in the latest [commit](https://github.com/openvinotoolkit/openvino/pull/1469/commits/42def620158ec449240f0e8d282090346a48ea81). What do you think about such implementation?